--- telnet/commands.c	2011-07-25 10:11:08.430863890 +0200
+++ telnet/commands.c.oden	2011-07-25 10:12:19.474303055 +0200
@@ -2412,13 +2412,17 @@ tn(int argc, char *argv[])
 		    ahints.ai_flags = AI_PASSIVE;
 		    error = getaddrinfo(aliasp, "0", &ahints, &ares);
 		    if (error) {
+			printf ("Couldn't get address for %s\n", aliasp);
 			warn("%s: %s", aliasp, gai_strerror(error));
 			close(net);
+			net = -1;
 			continue;
 		    }
 		    if (bind(net, ares->ai_addr, ares->ai_addrlen) < 0) {
+			printf ("Couldn't bind to %s\n", aliasp);
 			perror(aliasp);
 			(void) close(net);   /* dump descriptor */
+			net = -1;
 			freeaddrinfo(ares);
 			continue;
 		    }
